### PR TITLE
Fix CUDA Multinomial checks

### LINKED ISSUE #3475: CUDA tensor allows negative values in torch.multinomial
```python
import torch
t = torch.rand(10) #create CPU tensor
t.multinomial(1) #returns a result, as expected
tn = -t
tn.multinomial(1) # throws an exception, as per the documentation
tn = tn.cuda()
tn.multinomial(1) # returns a value, don't know from what distribution
tn.multinomial(1000, True) # returns a tensor of 1000 zeros
```

Looking at the values returned from tn.multinomial(1) and counting them, it seems like it returns the values of the multinomial distribution taken from t, i.e. that it has ignored the signs of the values. However, it seem like `tn.multinomial(n where n > 1)` always returns a tensor filled with zeros.

### PR DESCRIPTION
**Pull request by ssnl**
Fixes #3475 .
Also adds probability non-negative check on CPU.

### CODE REVIEW

File: aten/src/THC/THCTensorRandom.cuh [hunk 1]
```diff
@@ -94,15 +94,19 @@ template <typename T>
...
-      sum = THCNumerics<T>::add(sum, dist[row * cols + col]);
+      val = dist[row * cols + col];
+      assert(THCNumerics<T>::ge(val, zero));
+      sum = THCNumerics<T>::add(sum, val);
     }
 
-    sum = reduceBlock(smem, blockDim.x, sum, ReduceAdd<T, T>(), ScalarConvert<int, T>::to(0));
+    sum = reduceBlock(smem, blockDim.x, sum, ReduceAdd<T, T>(), zero);
     if (threadIdx.x == 0) {
+      assert(THCNumerics<T>::gt(sum, zero));
```
Comment (killeent) [inline aten/src/THC/THCTensorRandom.cuh]: I'm not sure we need an assertion here. You are checking in the above loop that there are no negative values, so there should be no case that leads to a negative value from the reduction.

File: aten/src/THC/THCTensorRandom.cuh [hunk 2]
```diff
@@ -182,6 +187,7 @@ sampleMultinomialOnce(int64_t* dest,
     if (threadIdx.x == 0) {
       // Make sure the sum of our distribution didn't overflow
       assert(!isinf(sum));
+      assert(THCNumerics<AccT>::gt(sum, accZero));
```
Comment (killeent) [inline aten/src/THC/THCTensorRandom.cuh]: Same here.

File: aten/src/THC/THCTensorRandom.cuh (hunk 1 above)
Comment (SsnL) [reply aten/src/THC/THCTensorRandom.cuh]: They can be all zeros.

Comment (killeent) [reply aten/src/THC/THCTensorRandom.cuh]: Ah you are right, my b.

**Review by killeent — APPROVED** [review]
LGTM.

**Pull request merged**